homeassistant/components/media_player/androidtv.py

```diff
@@ -0,0 +1,421 @@
+
+        entity_id = service.data.get(ATTR_ENTITY_ID)
+        target_devices = [dev for dev in hass.data[DATA_KEY].values()
+                          if dev.entity_id in entity_id]
+
+        for target_device in target_devices:
+            target_device.do_action(params['action'])
+
+    def service_intent(service):
+        """Dispatch service calls to target entities."""
+        params = {key: value for key, value in service.data.items()
+                  if key != ATTR_ENTITY_ID}
+
+        entity_id = service.data.get(ATTR_ENTITY_ID)
+        target_devices = [dev for dev in hass.data[DATA_KEY].values()
+                          if dev.entity_id in entity_id]
+
+        for target_device in target_devices:
+            target_device.start_intent(params['intent'])
+
+    def service_key(service):
+        """Dispatch service calls to target entities."""
+        params = {key: value for key, value in service.data.items()
+                  if key != ATTR_ENTITY_ID}
+
+        entity_id = service.data.get(ATTR_ENTITY_ID)
+        target_devices = [dev for dev in hass.data[DATA_KEY].values()
+                          if dev.entity_id in entity_id]
+
+        for target_device in target_devices:
+            target_device.input_key(params['key'])
+
+    hass.services.register(
+        DOMAIN, ACTION_SERVICE, service_action, schema=SERVICE_ACTION_SCHEMA)
+    hass.services.register(
+        DOMAIN, INTENT_SERVICE, service_intent, schema=SERVICE_INTENT_SCHEMA)
+    hass.services.register(
+        DOMAIN, KEY_SERVICE, service_key, schema=SERVICE_KEY_SCHEMA)
+
+
+class AndroidTv(MediaPlayerDevice):
+    """Representation of an AndroidTv device."""
+
+    def __init__(self, name, uri, client, adb_device):
+        """Initialize the Android device."""
+        self._client = client
+        self._adb_device = adb_device
+        self._uri = uri
+        self._name = name
+        self._available = None
+        self._volume = None
+        self._muted = None
+        self._state = None
+        self._app_id = None
+        self._app_name = None
+        self._device = None
+
+    def update(self):
+        """Get the latest details from the device."""
+        try:
+            devices = self._client.devices()
+        except RuntimeError:
+            # Can't connect to adb server
+            if self._available:
+                _LOGGER.error(
+                    "Can't reach adb server, is it running ?")
+                _LOGGER.warning(
+                    "Device {} became unavailable.".format(self._name))
+                self._available = False
+        else:
+            if any([self._uri in host.get_serial_no() for host in devices]):
+                # Device is connected through the ADB server
+                if not self._available:
+                    _LOGGER.info("Device {} reconnected.".format(self._name))
+                    self._available = True
+
+                power_output = self._adb_device.shell('dumpsys power')
```

Comment:
You need to create a new PyPi package that depends on `pure-python-adb` and that abstracts all these raw commands away into Python classes.